Simperium/SPAuthenticationValidator.m

```diff
@@ -7,45 +7,111 @@
 //
 
 #import "SPAuthenticationValidator.h"
+#import "NSError+Simperium.h"
+#import "NSString+Simperium.h"
 
-static int kDefaultMinimumPasswordLength = 4;
+static NSInteger SPAuthenticationDefaultMinPasswordLength = 8;
+static NSInteger SPAuthenticationLegacyMinPasswordLength = 4;
+
+NSString* const SPAuthenticationErrorDomain = @"SPAuthenticationValidatorDomain";
 
 @implementation SPAuthenticationValidator
-@synthesize minimumPasswordLength;
 
 - (instancetype)init {
     self = [super init];
     if (self) {
-        self.minimumPasswordLength = kDefaultMinimumPasswordLength;
+        self.strongMinimumPasswordLength = SPAuthenticationDefaultMinPasswordLength;
+        self.legacyMinimumPasswordLength = SPAuthenticationLegacyMinPasswordLength;
     }
     
     return self;
 }
 
 - (BOOL)isValidEmail:(NSString *)checkString {
     // From http://stackoverflow.com/a/3638271/1379066
-    BOOL stricterFilter = NO; // Discussion http://blog.logichigh.com/2010/09/02/validating-an-e-mail-address/
-    NSString *stricterFilterString = @"[A-Z0-9a-z\\._%+-]+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2,4}";
-    NSString *laxString = @".+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2}[A-Za-z]*";
-    NSString *emailRegex = stricterFilter ? stricterFilterString : laxString;
+    NSString *emailRegex = @".+@([A-Za-z0-9-]+\\.)+[A-Za-z]{2}[A-Za-z]*";
     NSPredicate *emailTest = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", emailRegex];
     
     return [emailTest evaluateWithObject:checkString];
 }
 
-- (BOOL)validateUsername:(NSString *)username {
-    // Expect email addresses by default
-    return [self isValidEmail:username];
+- (BOOL)validateUsername:(NSString *)username error:(NSError **)error {
+    if ([self isValidEmail:username]) {
+        return YES;
+    }
+
+    if (error) {
+        NSString *description = NSLocalizedString(@"Not a valid email address", @"Error when you enter a bad email address");
+
+        *error = [NSError sp_errorWithDomain:SPAuthenticationErrorDomain
+                                        code:SPAuthenticationErrorsEmailInvalid
+                                 description:description];
+    }
+
+    return NO;
 }
 
-- (BOOL)validatePasswordSecurity:(NSString *)password {
-    if ([password length] < self.minimumPasswordLength) {
+- (BOOL)validatePasswordWithUsername:(NSString *)username password:(NSString *)password error:(NSError **)error {
+    if (password.length < self.strongMinimumPasswordLength) {
+        if (error) {
+            NSString *description = NSLocalizedString(@"Password must contain at least %d characters", comment: @"Message displayed when password is too short. Please preserve the Percent D!");
```

Comment:
`Please preserve the Percent D!`
I like it, but would it be more clear to our polyglots to say something like `The "%d" is a placeholder for a numeral. Please preserve the Percent D!`

Reply:
Done, thank you!!